### ISSUE TITLE
[Feature request] Support functional in parameter of addAfter/addBefore method

### ISSUE THREAD

**Issue by oanhnn** [preset/preset#52]
I want add an environment variable base existed environment variable in `.env` file.
I was try with `replace().with()`, but ~~it don't work.~~ (See below comments)

```js
module.exports = Preset.make('Update ENV')
  .edit('.env')
  .replace(/^APP_URL=http:\/\/(.*)$/)
  .with((str, domain) => {
    return ['APP_URL=http://' + domain, 'APP_DOMAIN=' + domain].join('\n')
  })
  .chain()
```

I think `addAfter`/`addBefore` method should support functional like `replace().with()`.

```js
module.exports = Preset.make('Update ENV')
  .edit('.env')
  .search(/^APP_URL=http:\/\/(.*)$/)
    .addAfter((str, domain) => 'APP_DOMAIN=' + domain)
    .end()
  .chain()
```

**Comment by innocenzi** [comment]
What do you mean by "it doesn't work"? Is it an issue with the `\n`? 

This issue brings to the table the fact that it would be nice to be able to get information with Preset. In your case, you'd like to know the value of `APP_URL` thanks to a RegExp. This is something that could be stored in the preset context and used after, in this case with `addAfter`. 

Of course, we could also implement what you want, but it would probably imply a lot of changes in the current not-so-pretty code. 

When I've time to work again on implementing functionalities, I'll see what is the best option here and update this issue accordingly. 👍

**Comment by oanhnn** [comment]
### I setup test as below

```shell
$ mkdir -p preset-test/test
$ cd preset-test
$ npx use-preset preset
$ vi test/.env
$ tree -L 2 -I node_modules -a .
.
├── .gitignore
├── package.json
├── README.md
├── src
│   └── preset.js
├── templates
│   └── helloworld.txt
├── test
│   └── .env
└── yarn.lock

3 directories, 7 files
```

Content of `.env` file 

```
APP_NAME=Laravel
APP_URL=http://example-laravel.test
LOG_CHANNEL=stack
```

Edit `preset.js`  and run test:

```shell
$ cd test
$ npx npx use-preset ..
```

### Case 1

```js
const { Preset } = require('use-preset');

module.exports = Preset.make('test')
  .edit('.env')
  .replace(/^APP_URL=http:\/\/(.*)$/)
  .with((str, domain) => {
    return ['APP_URL=http://' + domain, 'APP_DOMAIN=' + domain].join('\n');
  })
  .chain();
```

**Result**

File `.env` don't change

### Case 2

```js
const { Preset } = require('use-preset');

module.exports = Preset.make('test')
  .edit('.env')
  .replace(/APP_URL=http:\/\/(.*)/)
  .with((str, domain) => {
    return ['APP_URL=http://' + domain, 'APP_DOMAIN=' + domain].join('\n');
  })
  .chain();
```

**Result**

```
APP_NAME=Laravel
APP_URL=http://undefined
APP_DOMAIN=undefined
LOG_CHANNEL=stack
```

### Case 3

```js
const { Preset } = require('use-preset');

module.exports = Preset.make('test')
  .edit('.env')
  .replace(/APP_URL=http:\/\/(.*)/)
  .with((str, domain) => {
    console.error(arguments)
    return ['APP_URL=http://' + str, 'APP_DOMAIN=' + domain].join('\n');
  })
  .chain();
```

**Result**

```
APP_NAME=Laravel
APP_URL=http://[object Object]
APP_DOMAIN=undefined
LOG_CHANNEL=stack
```
Console don't log error (list arguments)

### Case 4


```js
const { Preset } = require('use-preset');

module.exports = Preset.make('test')
  .edit('.env')
  .replace(/APP_URL=http:\/\/(.*)/)
  .with(function () {
	console.error(arguments)
    return 'APP_URL=http://abc.xyz\nAPP_DOMAIN=abc.xyz';
  })
  .chain();
```

**Result**

```
APP_NAME=Laravel
APP_URL=http://abc.xyz
APP_DOMAIN=abc.xyz
LOG_CHANNEL=stack
```

Console log error (list arguments)

```shell
$ npx use-preset ..
✔ Resolve
✔ Parse
❯ Apply test
  ❯ Edit file
    ✔ Validate
    ⠋ Execute
◼ Clean up temporary files
[Arguments] {
  '0': {
    generator: {
      name: 'test',
      templates: undefined,
      after: undefined,
      afterEach: undefined,
      before: undefined,
      beforeEach: undefined,
      options: [],
      actions: [Function: actions]
    },
    debug: false,
    targetDirectory: '/home/oanhnn/github/oanhnn/preset-test/test',
    task: TaskWrapper { task: [Task], errors: [], options: [Object] },
    argv: [ '..' ],
    temporary: false,
    presetName: 'test',
    presetDirectory: '/home/oanhnn/github/oanhnn/preset-test',
    presetTemplates: '/home/oanhnn/github/oanhnn/preset-test/templates',
✔ Resolve
✔ Parse
✔ Apply test
  ✔ Edit file
↓ Delete temporary directory [SKIPPED]
```

### Reason

1. My problem is that the regular expression was not correct
2. It working with "\n" character
3. Arguments list of replace function don't like reference *A string or a [replacement function.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace)*

   Arguments list should like: 

   ```js
   function (match, p1, p2, ..., offset, string, groups) {
      // ...
   }
   ```

   see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace#Specifying_a_function_as_a_parameter

4. Method `addAfter` and `addBefore` should support a function as parameter

**Thankyou**

**Comment by innocenzi** [comment]
In `0.2`, you'll be able to directly edit an environment file. Something like the following.

```js
Preset.setEnv('APP_NAME', 'Laravel', '.env'); // last is optional, defaults to .env
Preset.setEnv('APP_URL', ({ env }) => `https://${env.APP_DOMAIN}`);
```

The `edit` action will have fewer features because it wasn't maintainable, and the API drastically changes anyway between `0.1` and `0.2`, so I'll close this issue.

**EDIT**: the following API is implemented in `0.2`.
```js
Preset.env('.env') // optional, .env by default
	.set('APP_NAME', ({ APP_NAME }) => APP_NAME ?? 'Laravel')
	.set('APP_URL', ({ APP_NAME }) => `https://${APP_NAME.toLowerCase()}.test`)
	.skipIfMissing(); // by default, it's created if it's missing
```
